### ISSUE TITLE
line wrap in html breaks `numberLines` code line numbering

### ISSUE THREAD

**Issue by dbaynard** [jgm/pandoc#3838]
When printing html it is useful to wrap lines of code; however the supported option for adding line numbers leads to issues such as the following:

> ``` {.haskell .numberLines}
> extraLongFunctionWithVeryLongNames :: a -> a
> extraLongFunctionWithVeryLongNames = whenTheLineGetsTooLong . itWraps . whenPrinted .  ifTheCSSOption "white-space" . isSetTo "pre-wrap"
> 
> whenTheLineGetsTooLong :: a -> a
> whenTheLineGetsTooLong = id
> ```
> 
> becomes
> 
> ![line-numbers-misaligned](https://user-images.githubusercontent.com/870937/29004171-3a6f6d9c-7abb-11e7-9bbb-0edea8a2a468.png)

The fundamental issue is the markdown is converted to the following html

```html
<body>
<div class="sourceCode"><table class="sourceCode haskell numberLines"><tr class="sourceCode"><td class="lineNumbers"><pre>1
2
3
4
5
</pre></td>
<td class="sourceCode"><pre><code class="sourceCode haskell">
<span class="ot">extraLongFunctionWithVeryLongNames ::</span> a <span class="ot">-&gt;</span> a
extraLongFunctionWithVeryLongNames <span class="fu">=</span> whenTheLineGetsTooLong <span class="fu">.</span> itWraps <span class="fu">.</span> whenPrinted <span class="fu">.</span> ifTheCSSOption <span class="st">&quot;white-space&quot;</span> <span class="fu">.</span> isSetTo <span class="st">&quot;pre-wrap&quot;</span>

<span class="ot">whenTheLineGetsTooLong ::</span> a <span class="ot">-&gt;</span> a
whenTheLineGetsTooLong <span class="fu">=</span> id
</code></pre></td>
</tr></table></div>
</body>
```

With `numberLines` the code is inserted into a one row, two column table, with the left column a simple list of newline-separated line numbers. 
Without `numberLines` there is no indication what code corresponds to a line, in the html output — indeed, in the native AST line breaks are represented as `\n` but this information is lost on conversion to html.

I suspect the fix for this would be to rewrite the html writer so that `numberLines` wraps each line in a div, say (this could then be prefixed with a line number using css) or something more similar to github, where the code is presented in a table but each line has its own row.

What is the best approach?

---

Using
```
pandoc 1.19.2.1
Compiled with pandoc-types 1.17.0.5, texmath 0.9.4, skylighting 0.3.3
```

**Comment by dbaynard** [comment]
I should add:

-   this issue appears to relate to `skylightling` but it isn't clear where a fix might best go
-   this issue does not depend on language highlighting

**Comment by jgm** [comment]
We wanted to make sure that the code in a numbered code
block can be cut-and-pasted without getting the numbers
too.  That ruled out using a table, and I think it rules
out putting each line in its own div as well.

**Comment by bucklereed** [comment]
Copy-and-paste behaviour is a bit under-specified, but pseudo elements could fit the bill. Wrap each line in a span (or a div) with a class on each ('line', say), and then have this (untested) CSS:

```
.sourceCode { counter-reset: line 1 }
.line::before { content: counter(line); counter-increment: line }
```

**Comment by jgm** [comment]
In any case, this should be an issue on jgm/skylighting, since all of this is done by that library.

**Comment by dbaynard** [comment]
OK, I'll raise an issue there; I'll leave this open as there may be some pandoc decisions once the issue is fixed upstream.

**Comment by jgm** [comment]
This is closed by #3905
